File: projects/basis/basis/modules/market_reader.py
from web3 import Web3
from .factory import load_abi


class MarketReaderModule:
    def __init__(self, client, reader_address: str):
        self.client = client
        self.reader_address = Web3.to_checksum_address(reader_address)
        self.reader_abi = load_abi('AMarketReader.json')
        self.contract = self.client.web3.eth.contract(address=self.reader_address, abi=self.reader_abi)

    # ------------------------------------------------------------------
    # Read methods
    # ------------------------------------------------------------------

    def get_all_outcomes(self, router_address: str, market_token: str) -> list:
        """Returns all outcomes for a market."""
        checksum_router = Web3.to_checksum_address(router_address)
        checksum_market = Web3.to_checksum_address(market_token)
        return self.contract.functions.getAllOutcomes(checksum_router, checksum_market).call()

    def estimate_shares_out(self, router_address: str, market_token: str, outcome_id: int, usdb_amount: int, order_ids: list[int], user: str) -> int:
        """Estimates the number of shares received for a given USDB input."""
        checksum_router = Web3.to_checksum_address(router_address)
        checksum_market = Web3.to_checksum_address(market_token)
        checksum_user = Web3.to_checksum_address(user)
        return self.contract.functions.estimateSharesOut(checksum_router, checksum_market, outcome_id, usdb_amount, order_ids, checksum_user).call()

    def get_potential_payout(self, router_address: str, market_token: str, outcome_id: int, shares_amount: int, estimated_usdb_to_pool: int) -> tuple:
        """Returns the potential payout for selling shares."""
        checksum_router = Web3.to_checksum_address(router_address)
        checksum_market = Web3.to_checksum_address(market_token)
        return self.contract.functions.getPotentialPayout(checksum_router, checksum_market, outcome_id, shares_amount, estimated_usdb_to_pool).call()

File: projects/basis/src/modules/MarketReader.ts
import { BasisClient } from '../BasisClient';
import AMarketReaderArtifact from '../abis/AMarketReader.json';
import { Address } from 'viem';

export class MarketReaderModule {
  private client: BasisClient;
  private readerAddress: Address;

  constructor(client: BasisClient, readerAddress: Address) {
    this.client = client;
    this.readerAddress = readerAddress;
  }

  /**
   * Returns outcome info for all outcomes in a market.
   */
  async getAllOutcomes(routerAddress: Address, marketToken: Address) {
    return this.client.publicClient.readContract({
      address: this.readerAddress,
      abi: AMarketReaderArtifact.abi,
      functionName: 'getAllOutcomes',
      args: [routerAddress, marketToken],
    });
  }

  /**
   * Estimates the number of shares received for a given USDB input,
   * considering both order book fills and AMM.
   */
  async estimateSharesOut(
    routerAddress: Address,
    marketToken: Address,
    outcomeId: number,
    usdbAmount: bigint,
    orderIds: bigint[],
    user: Address
  ): Promise<bigint> {
    return this.client.publicClient.readContract({
      address: this.readerAddress,
      abi: AMarketReaderArtifact.abi,
      functionName: 'estimateSharesOut',
      args: [routerAddress, marketToken, outcomeId, usdbAmount, orderIds, user],
    }) as Promise<bigint>;
  }

  /**
   * Returns potential payout for holding or selling shares.
   */
  async getPotentialPayout(
    routerAddress: Address,
    marketToken: Address,
    outcomeId: number,
    sharesAmount: bigint,
    estimatedUsdbToPool: bigint
  ) {
    const result = await this.client.publicClient.readContract({
      address: this.readerAddress,
      abi: AMarketReaderArtifact.abi,
      functionName: 'getPotentialPayout',
      args: [routerAddress, marketToken, outcomeId, sharesAmount, estimatedUsdbToPool],
    }) as [bigint, bigint];

    return {
      holdPayout: result[0],
      simulatedAmmPayout: result[1],
    };
  }
}
